trunk/source/Library-ModelCheckerUtils/src/de/uni_freiburg/informatik/ultimate/lib/modelcheckerutils/cfg/IcfgPetrifier.java

```diff
@@ -52,98 +52,80 @@
-						addThreadInUseViolationVariablesAndErrorLocation, numberOfThreadInstances);
+		final Map<IIcfgForkTransitionThreadCurrent<IcfgLocation>, List<ThreadInstance>> threadInstanceMap =
+				ThreadInstanceAdder.constructThreadInstances(mPetrifiedIcfg, newForkCurrentThreads,
+						numberOfThreadInstances);
 		final Map<IIcfgForkTransitionThreadCurrent<IcfgLocation>, IcfgLocation> inUseErrorNodeMap = new HashMap<>();
 		final CfgSmtToolkit cfgSmtToolkit = adder.constructNewToolkit(mPetrifiedIcfg.getCfgSmtToolkit(),
-				threadInstanceMap, inUseErrorNodeMap, newJoinCurrentThreads, addThreadInUseViolationVariablesAndErrorLocation);
-		((BasicIcfg<IcfgLocation>) mPetrifiedIcfg).setCfgSmtToolkit(cfgSmtToolkit);
-		final HashRelation<String, String> copyDirectives = ProcedureMultiplier
-				.generateCopyDirectives(getAllInstances(threadInstanceMap));
+				threadInstanceMap, inUseErrorNodeMap, newJoinCurrentThreads);
+		mPetrifiedIcfg.setCfgSmtToolkit(cfgSmtToolkit);
+		final List<ThreadInstance> instances = getAllInstances(threadInstanceMap);
 		// Note that threadInstanceMap, newForkCurrentThreads, and
 		// newJoinCurrentThreads are modified because the
 		// ProcedureMultiplier might introduce new
 		// IcfgForkThreadCurrentTransitions, namely in the case where
 		// a forked transition contains a fork.
-		new ProcedureMultiplier(mServices, (BasicIcfg<IcfgLocation>) mPetrifiedIcfg, copyDirectives, backtranslator,
-				threadInstanceMap, newForkCurrentThreads, newJoinCurrentThreads);
+		ProcedureMultiplier.duplicateProcedures(mServices, mPetrifiedIcfg, instances, backtranslator, threadInstanceMap,
+				newForkCurrentThreads, newJoinCurrentThreads, this);
 		fillErrorNodeMap(threadInstanceMap.keySet(), inUseErrorNodeMap);
 		for (final Entry<IIcfgForkTransitionThreadCurrent<IcfgLocation>, IcfgLocation> entry : inUseErrorNodeMap
 				.entrySet()) {
-			((BasicIcfg<IcfgLocation>) mPetrifiedIcfg).addLocation(entry.getValue(), false, true, false, false, false);
+			mPetrifiedIcfg.addLocation(entry.getValue(), false, true, false, false, false);
 		}
-//		if (icfgConstructionMode == IcfgConstructionMode.CHECK_THREAD_INSTANCE_SUFFICIENCY) {
-//			ThreadInstanceAdder.addInUseErrorLocations((BasicIcfg<IcfgLocation>) mPetrifiedIcfg,
-//					getAllInstances(threadInstanceMap));
-//		}
 		final boolean addThreadInUseViolationEdges = true;
-				//(icfgConstructionMode == IcfgConstructionMode.CHECK_THREAD_INSTANCE_SUFFICIENCY);
 		adder.connectThreadInstances(mPetrifiedIcfg, newForkCurrentThreads, newJoinCurrentThreads, threadInstanceMap,
 				inUseErrorNodeMap, backtranslator, addThreadInUseViolationEdges);
 
-		final Set<Term> auxiliaryThreadVariables = collectAxiliaryThreadVariables(getAllInstances(threadInstanceMap),
-				addThreadInUseViolationEdges);
-		backtranslator.setVariableBlacklist(auxiliaryThreadVariables);
+		backtranslator.setVariableBlacklist(collectAxiliaryThreadVariables(instances, addThreadInUseViolationEdges));

```

Comment:
Could you fix the typo in the method name?

Reply:
Even nicer, I made it a stream instead

Reply:
Ok, that resolves the typo I guess.